### ISSUE TITLE
What needs to be altered to add a new driver?

### ISSUE THREAD

**Issue by backXslash** [apollo-ng/governess#30]
I'm trying to add a driver for the MAX31856 on the Raspberry Pi. There are working drivers [here](https://github.com/johnrbnsn/Adafruit_Python_MAX31856), I'm just not sure what needs to be altered to get them to load alongside the MAX31855 drivers. 

Can someone point me in the right direction?

**Comment by chron0** [comment]
I'm not sure if the IRC web gateway somehow disconnected you while idling:

```
( 00:03  backXslas ) Is anyone able to point me in the right direction for adding a driver to Governess?
( 00:03  backXslas ) I screwed up and ordered a MAX31856 breakout instead of the 55, it's actually a better chip and Adafruit even has an RPi library for it 
( 00:20 backXslash x [43bfb8ce@gateway/web/freenode/ip.67.191.184.206] has quit irc [Quit: Page closed]
( 00:35  M-Dzzzzzz ) Hey backXslash, you'll want to wait for chron0, he's on European time, so may be in bed?
```

Luckily, your message wasn't lost and directly lead to https://github.com/apollo-ng/governess/commit/8f3400ee8bf3596ecd7911e14418b00b5a7d868f and https://github.com/apollo-ng/governess/commit/a9c4be658c1fe3272137745aa72b5c7dfdab83ce

![localhost-8100- galaxy s iii](https://cloud.githubusercontent.com/assets/3270028/22235802/973d4400-e1f9-11e6-9a4f-d516a99749f7.png)

It will still take some time until the actual server part is doing something useful though, for now it's only there not to be forgotten and to have a working code sample in place that can be used to refactor it to however the actual plugin code will look when it's finished.

If you consider yourself a python dev, we could speed things up by collaborating and starting to do the heavy lifting work to get the server part (#6) up and running :) If not, you'll have to wait a little while longer until governess can be actually used as a whole, working system.

**Comment by backXslash** [comment]
Awesome! And yeah, it did boot me or something, I didn't see anything, so I
left.

The MAX31856 drivers only support hardware SPI at the moment, so.... Heads
up there.

I'm not exactly a python God or anything, but I dabble. What's missing? How
can I help? There's a lot about governess that seems scattered, that's why
I asked about where to start.

**Comment by chron0** [comment]
Yeah, I can see how it must look from the outside, that's why I haven't done much so far to attract more devs since it would require to create many more docs and conceptual presentation in order to give people a picture of what I've got in my head. I'm not a god in anything as well, but picoReflow showed that there is a strong demand for a software stack like governess, so I started with what I could. As I'm even less python experienced and still had doubts about my capabilities to produce something intuitive and usable while maintaining freedom and flexibility I decided to put the cart before the horse and see if I could come up with a data-driven UI that would enable me to build the frontend workflows and would produce only data- and workflow-models, which would have to be implemented on the server side.

Ideally, have a look at #6, I've already added more info there to give you a better understanding of how things work. You could start to play with a totally simple approach in python, creating mock drivers/plugins which would build the JSON list for the client from single files within that directory structure on init. Afterwards we could add real hardware code and websocket stuff to get it to the client...

On the server side, there is no structure/definition or anything really working yet, I've only given it some dir structure that made sense to me where I could paste code libs/fragments we can start to (re)use. 

Once I've finished the current work on the appliance editor, I'll make a screencast with more details so that other people can pick up the specific names, what they mean and the concept behind it more quickly to get real collaboration going effectively.
